Add XCM Runtime APIs to Develop -> Interoperability (#336)
polkadot-developers/polkadot-docs: develop/interoperability/xcm-runtime-apis.md

- Access of a storage item
- Access of a bundle of related storage items
- Deriving a value from storage based on arguments

In this sction, you will learn about specific runtime APIs to support XCM processing and manipulation.

## Dry Run API

The [Dry-run API](https://paritytech.github.io/polkadot-sdk/master/xcm_runtime_apis/dry_run/index.html), given an extrinsic, or an XCM program, returns its effects:

- Execution result
- Local XCM (in the case of an extrinsic)
- Forwarded XCMs
- List of events

This API can be used on its own for dry-running purposes, for double-checking or testing, but it mainly shines when used in conjunction with the XcmPaymentApi, given that it works for estimating fees only if you know the specific XCM you want to execute or send.

### Dry Run Call

This API allows to dry-run any extrinsic and obtaint the outcome, if it fails or succeeds and also the local-xcm and remote xcm messages sent to other chains.

```rust

> --8<-- 'https://raw.githubusercontent.com/paritytech/polkadot-sdk/refs/heads/stable2412/polkadot/xcm/xcm-runtime-apis/src/dry_run.rs:67:67'

Author: I don't think it makes sense because the line number could change in different versions

    The extrinsic to be executed.

    ---

??? interface "Output parameters"

    ++"Result<CallDryRunEffects<Event>, Error>"++

    Effects of dry-running an extrinsic. If an error occurs, it is returned instead of the effects.

    ??? child "Type `CallDryRunEffects<Event>`"

        `execution_result` ++"DispatchResultWithPostInfo"++

        The result of executing the extrinsic.

        ---

        `emitted_events` ++"Vec<Event>"++

        The list of events fired by the extrinsic.

        ---

        `local_xcm` ++"Option<VersionedXcm<()>>"++

        The local XCM that was attempted to be executed, if any.

        ---

        `forwarded_xcms` ++"Vec<(VersionedLocation, Vec<VersionedXcm<()>>)>"++

        The list of XCMs that were queued for sending.

    ---

??? interface "Example"

    This example demonstrates how to simulate a cross-chain asset transfer from the Paseo network to the Pop Network using a reserve transfer mechanism. Instead of executing the actual transfer, the code shows how to test and verify the transaction's behavior through a dry run before performing it on the live network:

>     ***Usage with PAPI***

Author: This is part of the framework. I don't think we can do that. @eshaben

    ```js

>     --8<-- 'code/develop/interoperability/xcm-runtime-apis/call-example.js'

Author: The code snippets are actually js in the examples

>     --8<-- 'code/develop/interoperability/xcm-runtime-apis/call-example.js'

Author: We can hide certain sections, but then the copy button will only copy the visible code

```rust
--8<-- 'https://raw.githubusercontent.com/paritytech/polkadot-sdk/refs/heads/stable2412/polkadot/xcm/xcm-runtime-apis/src/dry_run.rs:70:70'
```

??? interface "Input parameters"

    `origin_location` ++"VersionedLocation"++ <span class="required" markdown>++"required"++</span>

    The location of the origin that will execute the xcm message.

    ---

    `xcm` ++"VersionedXcm<Call>"++ <span class="required" markdown>++"required"++</span>

    A versioned XCM message.

    ---

??? interface "Output parameters"

    ++"Result<XcmDryRunEffects<Event>, Error>"++

    Effects of dry-running an extrinsic. If an error occurs, it is returned instead of the effects.

    ??? child "Type `CallDryRunEffects<Event>`"

        `execution_result` ++"DispatchResultWithPostInfo"++

        The result of executing the extrinsic.

        ---

        `emitted_events` ++"Vec<Event>"++

        The list of events fired by the extrinsic.

        ---

        `forwarded_xcms` ++"Vec<(VersionedLocation, Vec<VersionedXcm<()>>)>"++

        The list of XCMs that were queued for sending.

    ---

??? interface "Example"

>     This example demonstrates how to simulate a teleport asset transfer from the Paseo network to the Asset Hub parachain. The code shows how to test and verify the received XCM message's behavior in the destination chain through a dry run on the live network.

Author: Agree, the first one is better

     ***Usage with PAPI***

    ```js
    --8<-- 'code/develop/interoperability/xcm-runtime-apis/xcm-example.js'
    ```

    ***Output***

    --8<-- 'code/develop/interoperability/xcm-runtime-apis/xcm-example-output.html'

    ---

## XCM Payment API

The XCM Payment API provides a standardized way to determine the costs and payment options for executing XCM messages. Specifically, it enables clients to:

- Retrieve the weight required to execute an XCM message
- Obtain a list of acceptable AssetIds for paying execution fees
- Calculate the cost of the weight in a specified AssetId
- Estimate the fees for XCM message delivery

This API eliminates the need for clients to guess execution fees or identify acceptable assets manually. Instead, clients can query the list of supported asset IDs formatted according to the XCM version they understand. With this information, they can weigh the XCM program they intend to execute and convert the computed weight into its cost using one of the acceptable assets.

> To use the API effectively, the client must already know the XCM program to be executed and the chains involved in the program’s execution.

Reviewer: "XCM program" sounds a bit ambiguous, at least to me; maybe "XCM instructions"?

```rust
--8<-- 'https://raw.githubusercontent.com/paritytech/polkadot-sdk/refs/heads/stable2412/polkadot/xcm/xcm-runtime-apis/src/fees.rs:50:50'
```

??? interface "Input parameters"

    `message` ++"VersionedXcm<()>"++ <span class="required" markdown>++"required"++</span>
    
    A versioned XCM message whose execution weight is being queried.

    ---

??? interface "Output parameters"

    ++"Result<Weight, Error>"++
    
    The calculated weight required to execute the provided XCM message. If the calculation fails, an error is returned instead.

    ---

??? interface "Example"

    This example demonstrates how to calculate the weight needed to execute a teleport transfer from the Paseo network to the Asset Hub parachain using the XCM Payment API. The result shows the required weight in terms of reference time and proof size needed in the destination chain.

    ***Usage with PAPI***

    ```js
    --8<-- 'code/develop/interoperability/xcm-runtime-apis/query-xcm-weight.js'
    ```

    ***Output***

    --8<-- 'code/develop/interoperability/xcm-runtime-apis/query-xcm-weight-output.html'

    ---

### Query Weight to Asset Fee

> converts a given weight into the corresponding fee for a specified `AssetId`. It allows clients to determine the cost of execution in terms of the desired asset.

Reviewer: ```suggestion
Converts a given weight into the corresponding fee for a specified `AssetId`. It allows clients to determine the cost of execution in terms of the desired asset.
```

```rust
--8<-- 'https://raw.githubusercontent.com/paritytech/polkadot-sdk/refs/heads/stable2412/polkadot/xcm/xcm-runtime-apis/src/fees.rs:58:58'
```

??? interface "Input parameters"

    `weight` ++"Weight"++ <span class="required" markdown>++"required"++</span>
    
    The execution weight to be converted into a fee.

    ---

    `asset` ++"VersionedAssetId"++ <span class="required" markdown>++"required"++</span>
    
    The asset in which the fee will be calculated. This must be a versioned asset ID compatible with the runtime.

    ---

??? interface "Output parameters"

    ++"Result<u128, Error>"++
    

>     The fee needed to pay for the execution for the given AssetId.

Reviewer: ```suggestion
    The fee needed to pay for the execution for the given `AssetId.`
```

    ***Output***

    --8<-- 'code/develop/interoperability/xcm-runtime-apis/call-example-output.html::11'
                ...
    --8<-- 'code/develop/interoperability/xcm-runtime-apis/call-example-output.html:189'

    ---

### Dry Run XCM

This API allows to directly dry-run an xcm message instead of an extrinsic and check if it will execute succesfully and what other xcm messages will be forwarded to other chains.

```rust
--8<-- 'https://raw.githubusercontent.com/paritytech/polkadot-sdk/refs/heads/stable2412/polkadot/xcm/xcm-runtime-apis/src/dry_run.rs:70:70'
```

??? interface "Input parameters"

    `origin_location` ++"VersionedLocation"++ <span class="required" markdown>++"required"++</span>

    The location of the origin that will execute the xcm message.

    ---

    `xcm` ++"VersionedXcm<Call>"++ <span class="required" markdown>++"required"++</span>

    A versioned XCM message.

    ---

??? interface "Output parameters"

    ++"Result<XcmDryRunEffects<Event>, Error>"++

    Effects of dry-running an extrinsic. If an error occurs, it is returned instead of the effects.

>     ??? child "Type `CallDryRunEffects<Event>`"

Reviewer 2: Can we include the error as well? Although it's the same as with calls

        The list of XCMs that were queued for sending.

    ---

??? interface "Example"

    This example demonstrates how to simulate a teleport asset transfer from the Paseo network to the Paseo Asset Hub parachain. The code shows how to test and verify the received XCM message's behavior in the destination chain through a dry run on the live network.

     ***Usage with PAPI***

    ```js
    --8<-- 'code/develop/interoperability/xcm-runtime-apis/xcm-example.js'
    ```

    !!!note
        Make sure to replace `INSERT_USER_ADDRESS` with your SS58 address before running the script.

    ***Output***

    --8<-- 'code/develop/interoperability/xcm-runtime-apis/xcm-example-output.html'

    ---

## XCM Payment API

The [XCM Payment API](https://paritytech.github.io/polkadot-sdk/master/xcm_runtime_apis/fees/index.html){target=\_blank} provides a standardized way to determine the costs and payment options for executing XCM messages. Specifically, it enables clients to:

- Retrieve the weight required to execute an XCM message
- Obtain a list of acceptable `AssetIds` for paying execution fees
- Calculate the cost of the weight in a specified `AssetId`

> - Estimate the fees for XCM message delivery

Reviewer 2: These are the individual functions exposed by the API, but what I'd put is just two things:
- Getting execution fees
- Getting delivery fees

The first use case makes use of those first 3 functions, but that's more lower level

Reviewer 2: Also, the first time we mention "weight", can we link to the concept of weight? Maybe give it a little intro. It's not such a common concept that it's 2D and not just called "gas"

Author: The scope of this page is to document the existing APIs themselves. I'm working on a tutorial that will use them together, and this tutorial will be linked from this page.

This API eliminates the need for clients to guess execution fees or identify acceptable assets manually. Instead, clients can query the list of supported asset IDs formatted according to the XCM version they understand. With this information, they can weigh the XCM program they intend to execute and convert the computed weight into its cost using one of the acceptable assets.

To use the API effectively, the client must already know the XCM program to be executed and the chains involved in the program’s execution.

### Query Acceptable Payment Assets

Retrieves the list of assets that are acceptable for paying fees when using a specific XCM version

```rust
--8<-- 'https://raw.githubusercontent.com/paritytech/polkadot-sdk/refs/heads/stable2412/polkadot/xcm/xcm-runtime-apis/src/fees.rs:43:43'
```

??? interface "Input parameters"

    `xcm_version` ++"Version"++ <span class="required" markdown>++"required"++</span>

    Specifies the XCM version that will be used to send the XCM message.

    ---

> ??? interface "Output parameters"

Reviewer 2: Same comment about the error

```rust
--8<-- 'https://raw.githubusercontent.com/paritytech/polkadot-sdk/refs/heads/stable2412/polkadot/xcm/xcm-runtime-apis/src/fees.rs:58:58'
```

??? interface "Input parameters"

    `weight` ++"Weight"++ <span class="required" markdown>++"required"++</span>
    
    The execution weight to be converted into a fee.

    ---

    `asset` ++"VersionedAssetId"++ <span class="required" markdown>++"required"++</span>
    
    The asset in which the fee will be calculated. This must be a versioned asset ID compatible with the runtime.

    ---

??? interface "Output parameters"

    ++"Result<u128, Error>"++
    
    The fee needed to pay for the execution for the given `AssetId.`

    ---

??? interface "Example"

    This example demonstrates how to calculate the fee for a given execution weight using a specific versioned asset ID (PAS token) on Paseo Asset Hub.

    ***Usage with PAPI***

    ```js
    --8<-- 'code/develop/interoperability/xcm-runtime-apis/query-weight-to-fee.js'
    ```

    ***Output***

    --8<-- 'code/develop/interoperability/xcm-runtime-apis/query-weight-to-fee-output.html'

>     ---

Reviewer 2: I would definitely add an example after this that brings it all together.

We've made the error before of having "docs" be just a reference. That doesn't really tell the user how to use our tools.

Author: We will include a step-by-step explanation on a tutorial explaining how to use them together

### Query Delivery Fees

Retrieves the delivery fees for sending a specific XCM message to a designated destination. The fees are always returned in a specific asset defined by the destination chain.

```rust
--8<-- 'https://raw.githubusercontent.com/paritytech/polkadot-sdk/refs/heads/stable2412/polkadot/xcm/xcm-runtime-apis/src/fees.rs:68:68'
```

??? interface "Input parameters"

    `destination` ++"VersionedLocation"++ <span class="required" markdown>++"required"++</span>
    
    The target location where the message will be sent. Fees may vary depending on the destination, as different destinations often have unique fee structures and sender mechanisms.

    ---

    `message` ++"VersionedXcm<()>"++ <span class="required" markdown>++"required"++</span>
    
    The XCM message to be sent. The delivery fees are calculated based on the message's content and size, which can influence the cost.

    ---

??? interface "Output parameters"

    ++"Result<VersionedAssets, Error>"++
    
    The calculated delivery fees expressed in a specific asset supported by the destination chain. If an error occurs during the query, it returns an error instead.

    ---

> ??? interface "Example"

Reviewer 2: Again, this setup with "input", "output" and "example" is useful, but I think the most useful thing we could do is show how these functions work together.

I would show:
- the whole process of how to estimate XCM fees
- how the dry run api can be used in an app to let the user know if an action will be successful or not

Author: This will be done in the tutorial

        ---

        `emitted_events` ++"Vec<Event>"++

        The list of events fired by the extrinsic.

        ---

        `local_xcm` ++"Option<VersionedXcm<()>>"++

        The local XCM that was attempted to be executed, if any.

        ---

        `forwarded_xcms` ++"Vec<(VersionedLocation, Vec<VersionedXcm<()>>)>"++

        The list of XCMs that were queued for sending.

    ??? child "Type `Error`"

        Enum:

        - **`Unimplemented`** - an API part is unsupported
        - **`VersionedConversionFailed`** - converting a versioned data structure from one version to another failed

    ---

??? interface "Example"

    This example demonstrates how to simulate a cross-chain asset transfer from the Paseo network to the Pop Network using a [reserve transfer](https://wiki.polkadot.network/docs/learn/xcm/journey/transfers-reserve){target=\_blank} mechanism. Instead of executing the actual transfer, the code shows how to test and verify the transaction's behavior through a dry run before performing it on the live network:

    ***Usage with PAPI***

    ```js
    --8<-- 'code/develop/interoperability/xcm-runtime-apis/call-example.js'
    ```

>     !!!note

Reviewer 3: Please take this out of the note and add it to the copy above the code sample so the user sees it before they start working with the script. Thank you! 

"This example demonstrates how to simulate a cross-chain asset transfer from the Paseo network to the Pop Network using a [reserve transfer](https://wiki.polkadot.network/docs/learn/xcm/journey/transfers-reserve){target=\_blank} mechanism. Instead of executing the actual transfer, the code shows how to test and verify the transaction's behavior through a dry run before performing it on the live network. 

Replace `INSERT_USER_ADDRESS` with your SS58 address before running the script."

        Make sure to replace `INSERT_USER_ADDRESS` with your SS58 address before running the script.

    ***Output***

    --8<-- 'code/develop/interoperability/xcm-runtime-apis/call-example-output.html::11'
                ...
    --8<-- 'code/develop/interoperability/xcm-runtime-apis/call-example-output.html:189'

    ---

### Dry Run XCM

> This API allows to directly dry-run an xcm message instead of an extrinsic and check if it will execute succesfully and what other xcm messages will be forwarded to other chains.

Reviewer 3: ```suggestion
This API allows the direct dry-run of an xcm message instead of an extrinsic one, checks if it will execute successfully, and determines what other xcm messages will be forwarded to other chains.
```

    ??? child "Type `XcmDryRunEffects<Event>`"

        `execution_result` ++"DispatchResultWithPostInfo"++

        The result of executing the extrinsic.

        ---

        `emitted_events` ++"Vec<Event>"++

        The list of events fired by the extrinsic.

        ---

        `forwarded_xcms` ++"Vec<(VersionedLocation, Vec<VersionedXcm<()>>)>"++

        The list of XCMs that were queued for sending.

    ??? child "Type `Error`"

        Enum:

        - **`Unimplemented`** - an API part is unsupported
        - **`VersionedConversionFailed`** - converting a versioned data structure from one version to another failed

    ---

??? interface "Example"

    This example demonstrates how to simulate a [teleport asset transfer](https://wiki.polkadot.network/docs/learn/xcm/journey/transfers-teleport){target=\_blank} from the Paseo network to the Paseo Asset Hub parachain. The code shows how to test and verify the received XCM message's behavior in the destination chain through a dry run on the live network.

     ***Usage with PAPI***

    ```js
    --8<-- 'code/develop/interoperability/xcm-runtime-apis/xcm-example.js'
    ```

>     !!!note

Reviewer 3: Same suggestion re: take out of box and add to copy above code example. ty

        ---

        `proof_size` ++"u64"++

        The weight of storage space used by proof of validity.

        ---

    ??? child "Type `Error`"

        Enum:

        - **`Unimplemented`** - an API part is unsupported
        - **`VersionedConversionFailed`** - converting a versioned data structure from one version to another failed
        - **`WeightNotComputable`** - XCM message weight calculation failed
        - **`UnhandledXcmVersion`** - XCM version not able to be handled
        - **`AssetNotFound`** - the given asset is not handled as a fee asset
        - **`Unroutable`** - destination is known to be unroutable

    ---

??? interface "Example"

    This example demonstrates how to calculate the weight needed to execute a [teleport transfer](https://wiki.polkadot.network/docs/learn/xcm/journey/transfers-teleport){target=\_blank} from the Paseo network to the Paseo Asset Hub parachain using the XCM Payment API. The result shows the required weight in terms of reference time and proof size needed in the destination chain.

    ***Usage with PAPI***

    ```js
    --8<-- 'code/develop/interoperability/xcm-runtime-apis/query-xcm-weight.js'
    ```

>     !!!note

Reviewer 3: Same suggestion re: take out of box and add to copy above code example. ty

    
    The XCM message to be sent. The delivery fees are calculated based on the message's content and size, which can influence the cost.

    ---

??? interface "Output parameters"

    ++"Result<VersionedAssets, Error>"++
    
    The calculated delivery fees expressed in a specific asset supported by the destination chain. If an error occurs during the query, it returns an error instead.

    ??? child "Type `Error`"

        Enum:

        - **`Unimplemented`** - an API part is unsupported
        - **`VersionedConversionFailed`** - converting a versioned data structure from one version to another failed
        - **`WeightNotComputable`** - XCM message weight calculation failed
        - **`UnhandledXcmVersion`** - XCM version not able to be handled
        - **`AssetNotFound`** - the given asset is not handled as a fee asset
        - **`Unroutable`** - destination is known to be unroutable

    ---

??? interface "Example"

    This example demonstrates how to query the delivery fees for sending an XCM message from Paseo to Paseo Asset Hub.

    ***Usage with PAPI***

    ```js
    --8<-- 'code/develop/interoperability/xcm-runtime-apis/query-delivery-fees.js'
    ```

>     !!!note

Reviewer 3: Same suggestion re: take out of box and add to copy above code example. ty
